python/lancedb/embeddings/functions.py

```diff
@@ -0,0 +1,192 @@
+#  Copyright (c) 2023. LanceDB Developers
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import json
+from abc import ABC, abstractmethod
+from typing import List, Union
+
+import numpy as np
+from cachetools import cached
+from pydantic import BaseModel
+
+
+class EmbeddingFunctionRegistry:
+    """
+    This is a singleton class is used to register embedding functions
+    and fetch them by name. It also handles serializing and deserializing
+    """
+
+    def __init__(self):
+        self._functions = {}
+
+    def register(self, alias: str = None):
+        """
+        This creates a decorator that can be used to register
+        an EmbeddingFunctionModel.
+
+        Parameters
+        ----------
+        alias : str, optional
+            The alias is used to identify the function when
+            serializing/deserializing. If no alias is supplied,
+            then the class name is used.
+        """
+
+        # This is a decorator for a class that inherits from BaseModel
+        # It adds the class to the registry
+        def decorator(cls):
+            if not issubclass(cls, EmbeddingFunctionModel):
+                raise TypeError("Must be a subclass of EmbeddingFunctionModel")
+            cls.__function_alias__ = alias or cls.__name__
+            self._functions[alias or cls.__name__] = cls
```

Comment:
Should it warn or reject when overriding a key( if someone accidentally overrides it with another function, it'll break for the existing table).
Maybe either just warn
Or
Throw exception and ask for explicit "overwrite" if they actually want to overwrite?

Reply:
I may end up just removing the alias for now to simplify things. But good point. For now we can just raise an error on conflict.